### PR TITLE
Add fold_mutref

### PR DESCRIPTION
**Pull request by cramertj**
I've wanted this function a couple times in the past, so i thought I'd suggest adding it here. Feel free to bikeshed the name-- I also considered `fold_mut`.

### CODE REVIEW

**Comment by bluss** [comment]
Warning, I'm thinking aloud.. It's not bad:

```rust
let element_freq = iterable.fold_mut(HashMap::new(), |map, c| *map.entry(c).or_insert(0) += 1);
```

In a way it seems redundant with `for_each` (finally it's coming!). We have closure environments, so we don't really need to pass map, the reason for that is mostly the expression oriented way we want to code. And the expression oriented way is most natural with the existing fold, the by value version.

(Examples for [unfold 0.6.0](https://docs.rs/itertools/0.6.0/itertools/fn.unfold.html) and [unfold 0.6.2](https://docs.rs/itertools/0.6.2/itertools/fn.unfold.html) stabs at trying both of those.. the way it's expressed with closure captures is far superior for this case.)

**Comment by cramertj** [comment]
It is kind of redundant with `for_each`, agreed.

**Comment by cramertj** [comment]
Note that some of the nastiness comes from having to use `*` and `ref mut` to unwrap the multiple bits of state. That'll be a lot cleaner once match-ergo lands.

File: src/lib.rs [hunk 1]
```diff
@@ -1308,6 +1308,35 @@ pub trait Itertools : Iterator {
...
+    /// let s = (0..10).fold_mutref(State(0), |s_mut, x| s_mut.mutate(x));
+    /// assert_eq!(s, State(45));
+    /// ```
+    fn fold_mutref<ACC, F>(&mut self, start: ACC, mut f: F) -> ACC
+        where F: FnMut(&mut ACC, Self::Item)
+    {
+        self.fold(start, |mut acc, item| {
+            f(&mut acc, item);
+            acc
+        })
```
Comment (bluss) [inline src/lib.rs]: I think a `foreach` (or `fold((), |(), elt| ..)`) based implementation is best, it's truest to the method that we don't move around the acc value at all, if that for some reason is a big inconvenience or does not optimize out.

**Comment by bluss** [comment]
it's a bit similar to this method as well: [rayon's `for_each_with`](https://docs.rs/rayon/0.8.2/rayon/iter/trait.ParallelIterator.html#method.for_each_with), but there is nothing quite like it in rayon.

**Comment by bluss** [comment]
Ok, sorry for letting this wait for a bit. I think it seems this offers too little in terms of a new feature, that we don't want to include it. Let me know if you have other use cases or users that change the perspective on that.

**Comment by cramertj** [comment]
Seems fine to me!

**Pull request closed without merging**